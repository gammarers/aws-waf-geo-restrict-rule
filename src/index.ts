import * as wafv2 from 'aws-cdk-lib/aws-wafv2';

export interface RuleConfig {
  readonly priority: number;
  readonly ruleName?: string;
  readonly cloudWatchMetricsName?: string;
}

export interface WAFGeoRestrictRuleProps {
  readonly allowCountries: string[];
}

export class WAFGeoRestrictRule {

  constructor(private props: WAFGeoRestrictRuleProps) {}

  allowRule(config: RuleConfig): wafv2.CfnWebACL.RuleProperty {
    return {
      name: config.ruleName || 'allow-geo-rule',
      priority: config.priority,
      action: { allow: {} },
      statement: {
        geoMatchStatement: {
          countryCodes: this.props.allowCountries,
        },
      },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: config.cloudWatchMetricsName || 'AllowGeoMetric',
        sampledRequestsEnabled: true,
      },
    };
  }

  blockRule(config: RuleConfig): wafv2.CfnWebACL.RuleProperty {
    return {
      name: config.ruleName || 'block-geo-rule',
      priority: config.priority,
      action: { block: {} },
      statement: {
        notStatement: {
          statement: {
            geoMatchStatement: {
              countryCodes: this.props.allowCountries,
            },
          },
        },
      },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: config.cloudWatchMetricsName || 'BlockGeoMetric',
        sampledRequestsEnabled: true,
      },
    };
  }
}
