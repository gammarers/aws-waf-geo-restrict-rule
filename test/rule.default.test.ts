import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as wafv2 from 'aws-cdk-lib/aws-wafv2';
import { WAFGeoRestrictRule } from '../src';

describe('Web ACL Rule Group Global Scope default testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  const geoRestrictRule = new WAFGeoRestrictRule({
    allowCountries: ['JP'],
  });

  new wafv2.CfnWebACL(stack, 'WebACL', {
    defaultAction: { allow: {} },
    scope: 'CLOUD_FRONT',
    name: 'WebAclWithCustomRules',
    visibilityConfig: {
      cloudWatchMetricsEnabled: true,
      metricName: 'WebAclMetric',
      sampledRequestsEnabled: true,
    },
    rules: [
      geoRestrictRule.allowRule({
        priority: 1,
      }),
      geoRestrictRule.blockRule({
        priority: 2,
      }),
    ],
  });

  const template = Template.fromStack(stack);

  it('Should have WAF WebACL Rule', async () => {
    template.hasResourceProperties('AWS::WAFv2::WebACL', {
      Rules: [
        {
          Action: {
            Allow: {},
          },
          Name: 'allow-geo-rule',
          Priority: 1,
          Statement: {
            GeoMatchStatement: {
              CountryCodes: ['JP'],
            },
          },
          VisibilityConfig: {
            CloudWatchMetricsEnabled: true,
            MetricName: 'AllowGeoMetric',
            SampledRequestsEnabled: true,
          },
        },
        {
          Action: {
            Block: {},
          },
          Name: 'block-geo-rule',
          Priority: 2,
          Statement: {
            NotStatement: {
              Statement: {
                GeoMatchStatement: {
                  CountryCodes: ['JP'],
                },
              },
            },
          },
          VisibilityConfig: {
            CloudWatchMetricsEnabled: true,
            MetricName: 'BlockGeoMetric',
            SampledRequestsEnabled: true,
          },
        },
      ],
    });
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});