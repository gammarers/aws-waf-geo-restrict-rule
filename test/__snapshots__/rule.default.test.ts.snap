// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web ACL Rule Group Global Scope default testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "WebACL": {
      "Properties": {
        "DefaultAction": {
          "Allow": {},
        },
        "Name": "WebAclWithCustomRules",
        "Rules": [
          {
            "Action": {
              "Allow": {},
            },
            "Name": "allow-geo-rule",
            "Priority": 1,
            "Statement": {
              "GeoMatchStatement": {
                "CountryCodes": [
                  "JP",
                ],
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AllowGeoMetric",
              "SampledRequestsEnabled": true,
            },
          },
          {
            "Action": {
              "Block": {},
            },
            "Name": "block-geo-rule",
            "Priority": 2,
            "Statement": {
              "NotStatement": {
                "Statement": {
                  "GeoMatchStatement": {
                    "CountryCodes": [
                      "JP",
                    ],
                  },
                },
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "BlockGeoMetric",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "CLOUD_FRONT",
        "VisibilityConfig": {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "WebAclMetric",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::WebACL",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
